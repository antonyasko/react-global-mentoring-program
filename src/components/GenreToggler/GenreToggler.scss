@import '../../global.scss';

.genre-toggler {
  background-color: $dark-gray-color;

  &__button {
    height: 57px;
    padding: 15px;
    font-style: normal;
    font-weight: 500;
    font-size: $font-size-m;
    line-height: $line-height-m;
    text-transform: uppercase;
    color: $white-color;
    border-bottom: 4px solid $dark-gray-color;

    &.active {
      border-color: $dark-indian-red-color;
      z-index: $z-index-modal;
    }
  }
}
