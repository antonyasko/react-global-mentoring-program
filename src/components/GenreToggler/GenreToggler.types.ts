export interface IGenreTogglerProps {
  genres: string[];
}
