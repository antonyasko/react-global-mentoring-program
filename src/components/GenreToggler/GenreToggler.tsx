import React, { SyntheticEvent, memo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import { IGenreTogglerProps } from './GenreToggler.types';

import { showByGenreAction } from '../../store/actions/cardsFiltersAction';
import { IFiltersState } from '../../store/types';

import './GenreToggler.scss';

function GenreToggler({ genres }: IGenreTogglerProps): JSX.Element {
  const dispatch = useDispatch();
  const { showByGenre } = useSelector((state: IFiltersState) => state.filters);
  const navigate = useNavigate();

  function onButtonClick(e: SyntheticEvent<HTMLButtonElement>): void {
    const genre = (e.target as HTMLButtonElement).dataset.genre as string;
    dispatch(showByGenreAction(genre));
    navigate(`genre=${genre}`);
  }

  return (
    <div className="container genre-toggler">
      {genres.map((genre) => (
        <button
          type="button"
          className={`genre-toggler__button ${genre === showByGenre ? 'active' : ''}`}
          data-genre={genre}
          key={genre}
          onClick={onButtonClick}
        >
          {genre}
        </button>
      ))}
    </div>
  );
}

export default memo(GenreToggler);
