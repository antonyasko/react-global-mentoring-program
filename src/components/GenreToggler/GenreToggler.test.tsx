import React from 'react';
import * as ReactRedux from 'react-redux';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import GenreToggler from './GenreToggler';

const TEST_GENRE = 'test-genre';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(() => jest.fn()),
  useSelector: jest.fn(() => ({ showByGenre: 'all' })),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn,
}));

const genres = ['all', 'comedy', 'action', 'triller', TEST_GENRE];

describe('<GenreToggler />', () => {
  const TestGenreToggler = <GenreToggler genres={genres} />;

  it('should match snapshot', () => {
    const { container } = render(TestGenreToggler);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('All button should be active by default', () => {
    const { getByText } = render(TestGenreToggler);

    const allButton = getByText('all');

    expect(allButton).toHaveClass('active');
  });

  it('Test button should be active', () => {
    jest.spyOn(ReactRedux, 'useSelector').mockImplementation(() => ({ showByGenre: TEST_GENRE }));

    const { getByText } = render(TestGenreToggler);

    const testButton = getByText(TEST_GENRE);

    expect(testButton).toHaveClass('active');
  });
});
