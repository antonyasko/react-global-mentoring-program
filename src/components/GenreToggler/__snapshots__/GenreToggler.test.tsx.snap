// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenreToggler /> should match snapshot 1`] = `
<div
  class="container genre-toggler"
>
  <button
    class="genre-toggler__button active"
    data-genre="all"
    type="button"
  >
    all
  </button>
  <button
    class="genre-toggler__button "
    data-genre="comedy"
    type="button"
  >
    comedy
  </button>
  <button
    class="genre-toggler__button "
    data-genre="action"
    type="button"
  >
    action
  </button>
  <button
    class="genre-toggler__button "
    data-genre="triller"
    type="button"
  >
    triller
  </button>
  <button
    class="genre-toggler__button "
    data-genre="test-genre"
    type="button"
  >
    test-genre
  </button>
</div>
`;
